### ISSUE TITLE
GCNCMC doesn't accept any proposed moves during 12 hours of simulation wall time  

### ISSUE THREAD

**Issue by noahharrison64** [essex-lab/gcncmc-paper#1]
Hi,

I was interested by the GCNCMC/MD paper proposed and wanted to have a go at implementing it myself. After setting up the input files  I attempted to run my first simulation (using the simulation parameters specified in [this script](https://github.com/MobleyLab/water_benchmark_paper/blob/main/supplementary-documents/scripts/simulations/grand/equil_uvt1/equil-uvt1.py))

Over the course of about 15 hours the log messages suggest not a single move was accepted, and the simulation did not finish running i.e. the following lines were continually being written to the log file

>2024-02-09 11:55:44,444 - INFO: Move rejected due to water leaving the GCMC sphere

I wondered if I had set up my system incorrectly, so I re-ran the simulation using the standard gcmc sampler, and found 18 moves were accepted over the course of the simulation, suggesting to me the issue wasn't necessarily to do with the system setup, unless I've missed something important in the docs.

**The code for the Non-EQ sampling setup is below:**

```py
import sys
from openmm import *
from openmm.app import *
from openmm.unit import *
from openmmtools.integrators import BAOABIntegrator
from sys import stdout
import numpy as np
import mdtraj
import grand
from pathlib import Path
# Load PDB

input_dir = Path('./')
pdb = PDBFile((input_dir / 'em.pdb'))

# Add ghost waters
pdb.topology, pdb.positions, ghosts = grand.utils.add_ghosts(pdb.topology, pdb.positions,
                                                             n=15, pdb='non_eq-ghosts.pdb')

ff = ForceField('amber14/protein.ff14SB.xml', 'amber14/tip3p.xml', str(input_dir / 'ligand.xml'))

system = ff.createSystem(pdb.topology,
                         nonbondedMethod=PME,
                         nonbondedCutoff=10.0*angstroms,
                         switchDistance=8.0*angstroms,
                         constraints=HBonds)

sphere =  [47.12823245, 39.21931498, 43.33853147]

integrator = BAOABIntegrator(277*kelvin, 1.0/picosecond, 0.002*picosecond)

gcmc_mover = grand.samplers.NonequilibriumGCMCSphereSampler(system=system,
                                                              topology=pdb.topology,
                                                              temperature=300*kelvin,
                                                              integrator=integrator,
                                                              # Make this a 10 ps protocol
                                                              nPertSteps=99, nPropStepsPerPert=50,
                                                              sphereCentre = Quantity(np.array(sphere), angstroms),
                                                              sphereRadius=5.5*angstroms,
                                                              log='non_eq_gcmc.log',
                                                              dcd='raw.dcd',
                                                              rst='rst.rst7',
                                                              overwrite=True)

# Define integrator
# Define platform and set precision
platform = Platform.getPlatformByName('CUDA')
platform.setPropertyDefaultValue('Precision', 'mixed')

# Create simulation object
simulation = Simulation(pdb.topology, system, gcmc_mover.compound_integrator, platform)
# Set positions, velocities and box vectors
simulation.context.setPositions(pdb.positions)
simulation.context.setVelocitiesToTemperature(277*kelvin)
simulation.context.setPeriodicBoxVectors(*pdb.topology.getPeriodicBoxVectors())
```

**The code for the standard GCMC sampler is below:**

```py
# Load PDB
input_dir = Path('./')
pdb = PDBFile(str(input_dir / 'em.pdb'))

# Add ghost waters
pdb.topology, pdb.positions, ghosts = grand.utils.add_ghosts(pdb.topology, pdb.positions,
                                                             n=15, pdb='2xab-ghosts.pdb')

ff = ForceField('amber14/protein.ff14SB.xml', 'amber14/tip3p.xml', str(input_dir / 'ligand.xml'))

system = ff.createSystem(pdb.topology,
                         nonbondedMethod=PME,
                         nonbondedCutoff=10.0*angstroms,
                         switchDistance=8.0*angstroms,
                         constraints=HBonds)

sphere =  [47.12823245, 39.21931498, 43.33853147]

# Define GCMC Sampler
gcmc_mover = grand.samplers.StandardGCMCSphereSampler(system=system,
                                                      topology=pdb.topology,
                                                      temperature=277*kelvin,
                                                      # referenceAtoms=ref_atoms,
                                                      sphereCentre = Quantity(np.array(sphere), angstroms),
                                                      sphereRadius=5.5*angstroms,
                                                      log='2xab-equil-uvt1.log',
                                                      excessChemicalPotential=-6.34*kilocalorie_per_mole,
                                                      standardVolume=29.823*angstroms**3,
                                                      ghostFile='2xab-uvt1-ghosts.txt',
                                                      dcd='2xab-equil-uvt1.dcd',
                                                      rst='2xab-equil-uvt1.rst7',
                                                      overwrite=True)

# Define integrator
integrator = BAOABIntegrator(277*kelvin, 1.0/picosecond, 0.002*picosecond)
# Define platform and set precision
platform = Platform.getPlatformByName('CUDA')
platform.setPropertyDefaultValue('Precision', 'mixed')

# Create simulation object
simulation = Simulation(pdb.topology, system, integrator, platform)
# Set positions, velocities and box vectors
simulation.context.setPositions(pdb.positions)
simulation.context.setVelocitiesToTemperature(277*kelvin)
simulation.context.setPeriodicBoxVectors(*pdb.topology.getPeriodicBoxVectors())
```

**Code for running the simulation (either standard or Non-EQ sampling):**
```py
# Prepare the GCMC sphere
gcmc_mover.initialise(simulation.context, ghosts)
# Remove all waters currently in the sphere to reduce bias
gcmc_mover.deleteWatersInGCMCSphere()

# Start with 10k moves
#gcmc_mover.move(simulation.context, 10000)
print('Start with 10k moves')
gcmc_mover.move(simulation.context, 10000)


# Run GCMC/MD equilibration (100k GCMC moves over 1 ps - 1000 moves every 10 fs)
#for i in range(100):
for i in range(5):
    print(i)
    gcmc_mover.move(simulation.context, 1000)
    gcmc_mover.report(simulation)
    simulation.step(5)

# Remove ghosts and write out a PDB
# gcmc_ids = np.where(gcmc_mover.gcmc_status == 0)[0]
# ghost_resids = [gcmc_mover.gcmc_resids[id] for id in gcmc_ids]

positions = simulation.context.getState(getPositions=True, enforcePeriodicBox=True).getPositions()
pdb.topology, pdb.positions = grand.utils.remove_ghosts(pdb.topology, positions,
                                                        ghosts=ghosts,
                                                        pdb='2xab-uvt1.pdb')
```

I have attached the input files (energy_minimized.pdb and ligand.xml) as well as the log files from the non-eq sim (non_eq_gcmc.log) and the standard sim (2xab-equil-uvt1.log)

Any advice would be appreciated.

Thanks,
Noah

**Comment by olliemelling** [comment]
Hi, 

Happy to look into this for you, it looks like you may have forgotten to attach the input/output files?

If you could resend them then I'll take a look.

Thanks,

Ollie

**Comment by noahharrison64** [comment]
Oops uploaded to drive but forgot to post here! Thanks for having a look into it. 

[Here's the link](https://drive.google.com/drive/folders/1KO5eELXrfM_agdhonSF0kMS9SCE6Jn2r?usp=sharing)

**Comment by olliemelling** [comment]
Hi,

I think the issue here is to do with how you are defining the sphere - by using co-ordinates to define its centre and a radius to define its size. As the protein moves within the simulation box over the course of the simulation, the sphere centre does not get updated and therefore won't remain in the correct place.

If you define your sphere using a radius and set of reference atoms (example in the script you linked) then this should improve things. You can define any number of reference atoms and the sphere centre will just be taken as the midpoint of these reference atoms. Would recommend using alpha carbon, or at least backbone atoms as the reference atoms given their relative geometry is less likely to change.

Please let me know if you start to see some accepted moves.

Thanks,

Ollie

**Comment by noahharrison64** [comment]
Hi Ollie,

That makes sense. I went with the sphereCentre method just to make it more generalisable (easier to find binding site COM than an applicable set of atom names, I've found) Does that mean when the GCMC sphere is defined via a set of referenceAtoms the sphere centre is iteratively updated before the start of every move? Whereas if you supply sphereCentre as an argument then this is used statically throughout?

Thanks for looking into this, I'll give it a whirl using referenceAtoms instead!

Cheers,
Noah

**Comment by olliemelling** [comment]
Yep, that's correct - the sphere centre is only updated before each move when defined using reference atoms.

**Comment by noahharrison64** [comment]
Just gave this ago, an hour (update - 16 hours later still no moves accepted) into running the simulation and no moves have been accepted according to the logs...

I used the following code to generate the reference atoms:

```py
ref_atoms = [
    {
        'name': atom.type,
        'resname': atom.resname,
        'resid': str(atom.resid)
    } for atom in mda.Universe(str(input_dir / 'em.pdb')).select_atoms('around 5 resname MOL').atoms if atom.name == 'C'
]
```

which creates a list of dicts:

```
[{'name': 'C', 'resname': 'ALA', 'resid': '57'},
 {'name': 'C', 'resname': 'LEU', 'resid': '76'},
 {'name': 'C', 'resname': 'VAL', 'resid': '79'},
 {'name': 'C', 'resname': 'MET', 'resid': '80'},
 {'name': 'C', 'resname': 'VAL', 'resid': '83'},
 {'name': 'C', 'resname': 'ARG', 'resid': '93'},
 {'name': 'C', 'resname': 'THR', 'resid': '96'},
 {'name': 'C', 'resname': 'LEU', 'resid': '97'},
 {'name': 'C', 'resname': 'PHE', 'resid': '100'},
 {'name': 'C', 'resname': 'GLY', 'resid': '101'}]
```

and modified the init of the sampler as follows:

```py
gcmc_mover = grand.samplers.NonequilibriumGCMCSphereSampler(system=system,
                                                              topology=pdb.topology,
                                                              temperature=300*kelvin,
                                                              integrator=integrator,
                                                              # Make this a 10 ps protocol
                                                              nPertSteps=99, nPropStepsPerPert=50,
                                                              referenceAtoms=ref_atoms,
                                                              sphereRadius=5.5*angstroms,
                                                              log='gcmc.log',
                                                              dcd='raw.dcd',
                                                              rst='rst.rst7',
                                                              overwrite=True
)
```

Any further thoughts would be appreciated!

**Comment by olliemelling** [comment]
Hi,

I've had another look at this and visualised the sphere based on the reference atoms above. How many waters do you expect to be in the sphere during the simulation? There is only one water inside the sphere in the energy_minimized.pdb structure and this water is part of the bulk solvent so easy to see why it doesn't stay there.

It doesn't look like there's much space in the sphere for waters to be placed which may explain why you're not seeing any accepted moves.

Not sure if you use PyMol but I've attached a .pse file with the sphere visualised at 5.5 A radius and using your reference atoms. If you load this into pymol you should hopefully be able to see this.

It may be a case of making the sphere slightly larger depending on whereabouts you're expecting waters to be present in the binding site?

Thanks,

Ollie
[pymol.zip](https://github.com/essex-lab/gcncmc-paper/files/14310492/pymol.zip)

**Comment by noahharrison64** [comment]
Hi Ollie,

Thanks for continuing to look into this. I hear your point about the sphere size, but does that explain the discrepancy between GCMC/MD and GCMCNC/MD runtimes? Using the exact same protocol but simply swapping the samplers leads to a massive difference in run time. Non-EQ MC fails to complete, whereas the original GCMC/MD code takes about an hour. 

I tried increasing the sphere size to 8A however didn't really so an improvement - over an hour and a half of simulating no moves were accepted according to the log file. 

I'll have ago with a different protein ligand system next week and see if I see any moves accepted.

Cheers,
Noah

**Comment by olliemelling** [comment]
Hi,

Sorry I missed this point in your original message. It is expected for the non-eq method to take longer, but this often comes with higher acceptance rates (appreciate you aren't observing that here!).

Not sure how familiar you are with these grand canonical techniques and the differences between the two methods so apologies if this explanation doesn't answer your question... 

GCMC generates a trial state by simply placing a water at a random location/orientation within the sphere and uses the change in potential energy to determine whether that moves gets accepted. This is very quick to run and we can run 1000s of moves in a pretty short period of time, as you've seen.

GCNCMC generates a trial state by placing a water at a random location/orientation but then runs a short MD simulation (10 ps in your case) behind the scenes where the water's interactions are gradually strengthened to the point where it is fully coupled to the system. This allows the water to adapt to the environment and avoids any of the large steric clashes found with GCMC moves. But does come at an increased computational cost per move.

Although you are seeing moves accepted when using the GCMC sampler, these don't seem to be "useful" moves in that the number of waters in the sphere isn't changing - i.e. N still = 0 after the accepted move indicating that a water was just successfully inserted on the edge of the sphere and then diffused into the bulk solvent - not giving you any info about bound waters in your binding site.

As part of the acceptance criteria for GCNCMC, the moves are subject to an additional test in that the water must not leave the sphere during the 10 ps nonequilibrium simulation (this doesn't apply to GCMC as the GCMC moves are instantaneous and therefore the water can't leave). It seems like the same thing is happening in the GCNCMC simulations as in your GCMC simulations - the water gets inserted on the edge of the sphere and then diffuses into bulk solvent as it is coupled to the system.

In short, from the log files it seems like both methods are telling you the same thing - there are no hydration sites within your sphere.

Hopefully this makes some sense! More than happy to arrange a call to discuss this in more detail if that would be useful?

Thanks,

Ollie

**Comment by noahharrison64** [comment]
Ahh yes thank you for your explanation. I had done some reading on both your publication and the original GCMC paper but that was a while ago and it slipped my mind that with the NCMC there are perturbation steps in-between each MC set of moves. Of course that explains the difference in wall times between the two methodologies, apologies for the oversight on my part!

Thanks for explaining r.e. the non-useful moves we see in the GCMC run, that makes sense to me. I am not totally familiar with the system I'm using here, just pulled a random benchmarking target and didn't give much thought as to what the binding site looks like. 

I'll still have a look at both methods using a target with a different binding site, probably a kinase, just so I can get an idea of what these simulations produce when they actually have the option to swap water positions! Also thank you for the offer r.e. a call to discuss - I think your comments cleared up any misunderstandings on my part, although there's a chance I'll take you up on that offer if I get lost further into my pursuits of these GCMC sampling techniques!

Best,
Noah
